spring-cloud-gcp-logging/src/main/java/org/springframework/cloud/gcp/logging/TraceIdLoggingWebMvcInterceptor.java

```diff
@@ -0,0 +1,71 @@
+/*
+ *  Copyright 2017 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.logging;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.cloud.logging.TraceLoggingEnhancer;
+
+import org.springframework.web.servlet.handler.HandlerInterceptorAdapter;
+
+/**
+ * {@link org.springframework.web.servlet.HandlerInterceptor} that extracts the request
+ * trace ID from the "x-cloud-trace-context" HTTP header and stores it in a thread-local
+ * using {@link TraceLoggingEnhancer#setCurrentTraceId}.
+ *
+ * The {@link TraceLoggingEnhancer} can then be used in a logging appender to add the
+ * trace ID metadata to log messages.
+ *
+ * @author Mike Eltsufin
+ */
+public class TraceIdLoggingWebMvcInterceptor extends HandlerInterceptorAdapter {
+
+	private static final String X_CLOUD_TRACE = "x-cloud-trace-context";
+
+	@Override
+	public boolean preHandle(HttpServletRequest req,
+			HttpServletResponse resp, Object handler) throws Exception {
+		TraceLoggingEnhancer.setCurrentTraceId(extractTraceIdFromRequest(req));
+		return true;
+	}
+
+	@Override
+	public void afterCompletion(HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse, Object handler, Exception e) throws Exception {
+		TraceLoggingEnhancer.setCurrentTraceId(null);
```

Comment:
Yeah... That's not good. There is a `ThreadLocal.remove()` on the matter.
I mean that thread might be reused somewhere else, outside of the servlet container, but its state is polluted with a map with a `null` value.

Reply:
Agreed, but we're restricted by the `TraceLoggingEnhancer` API, which only has `setCurrentTraceId` and `getCurrentTraceId`.
This should be fixed upstream first.

Reply:
https://github.com/GoogleCloudPlatform/google-cloud-java/issues/2746